Data/HashMap/Internal.hs

```diff
@@ -1269,11 +1274,19 @@ update f = alter (>>= f)
 -- 'lookup' k ('alter' f k m) = f ('lookup' k m)
 -- @
 alter :: (Eq k, Hashable k) => (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
--- TODO(m-renaud): Consider using specialized insert and delete for alter.
 alter f k m =
-  case f (lookup k m) of
-    Nothing -> delete k m
-    Just v  -> insert k v m
+    let !h = hash k
+        !lookupRes = lookupRecordCollision h k m
+    in case f (lookupResToMaybe lookupRes) of
+        Nothing -> case lookupRes of
+            Absent            -> m
+            Present _ collPos -> deleteKeyExists collPos h k m
+        Just v' -> case lookupRes of
+            Absent            -> insertNewKey h k v' m
+            Present v collPos ->
+                if v `ptrEq` v'
+                    then m
+                    else insertKeyExists collPos h k v' m
```

Comment:
One bit of the generated STG for `$walter` that I thought was a bit weird is the allocation of a thunk that seems to contain the `Eq` methods for use in in `$wlookupRecordCollision#`:

```diff
diff --git a/old b/new
index 22b39f1..c8a222d 100644
--- a/old
+++ b/new
@@ -4,44 +4,41 @@ $walter [InlPrag=INLINABLE[2]]
      (Maybe v -> Maybe v) -> k -> HashMap k v -> HashMap k v
 [GblId,
  Arity=4,
- Str=<SP(LP(LCL(C1(L)),A),A,SCS(P(L)))><1C1(L)><SL><SL>,
+ Str=<SP(MP(LCL(C1(L)),A),A,1C1(P(L)))><1C1(L)><SL><SL>,
  Unf=OtherCon []] =
     {} \r [w w1 w2 w3]
+        case hash w w2 of {
+        I# i [Occ=Once1] ->
+        case int2Word# [i] of ipv {
+        __DEFAULT ->
         let {
-          sat [Occ=Once1] :: Maybe v
+          sat [Occ=Once1, Dmd=LP(LCL(C1(L)),A)] :: Eq k
           [LclId] =
-              {w, w2, w3} \u []
-                  case $wlookup# w w2 w3 of {
-                  (#,#) us us ->
-                  case us of us {
-                    __DEFAULT -> Nothing [];
-                    2# -> Just [us];
-                  };
-                  };
+              {w} \u [] $p1Hashable w;
         } in 
-          case w1 sat of {
-            Nothing ->
-                case hash w w2 of {
-                I# i [Occ=Once1] ->
-                case int2Word# [i] of sat [Occ=Once1] {
-                __DEFAULT ->
-                let {
-                  sat [Occ=Once1, Dmd=LP(LCL(C1(L)),A)] :: Eq k
-                  [LclId] =
-                      {w} \u [] $p1Hashable w;
-                } in  $wdelete' sat sat w2 w3;
-                };
+          case $wlookupRecordCollision# sat ipv w2 w3 of {
+          (#,,#) us us us ->
+          case us of us {
+            __DEFAULT ->
+                case w1 Nothing of {
+                  Nothing -> w3;
+                  Just v' [Occ=Once1] -> $winsertNewKey ipv w2 v' w3;
                 };
-            Just v1 [Occ=Once1] ->
-                case hash w w2 of {
-                I# i [Occ=Once1] ->
-                case int2Word# [i] of sat [Occ=Once1] {
-                __DEFAULT ->
+            2# ->
                 let {
-                  sat [Occ=Once1, Dmd=LP(LCL(C1(L)),A)] :: Eq k
+                  sat [Occ=Once1] :: Maybe v
                   [LclId] =
-                      {w} \u [] $p1Hashable w;
-                } in  $winsert' sat sat w2 v1 w3;
-                };
-                };
+                      CCCS Just! [us];
+                } in 
+                  case w1 sat of {
+                    Nothing -> $wdeleteKeyExists us ipv w2 w3;
+                    Just v' ->
+                        case reallyUnsafePtrEquality# [us v'] of {
+                          __DEFAULT -> $winsertKeyExists us ipv w2 v' w3;
+                          1# -> w3;
+                        };
+                  };
+          };
           };
+        };
+        };
```

I have no idea what we could do to avoid this though. I guess specialization should remove this overhead?!

Reply:
> I guess specialization should remove this overhead?!

Yeah, it does.